File: server/models/Player.js
import mongoose from "mongoose";

const defaultMaterials = {
    scales: 0,
    bones: 0,
    blood: 0,
    zima: 0,
    iride: 0,
    kobaureo: 0
};

const defaultPlants = {
    nillea: 0,
    tarmaret: 0,
    albalacea: 0,
    millis: 0,
    anthemon: 0,
    saelicornia: 0
};

const defaultElements = {
        fire: 0,
            horn: 0,
                coral: 0,
                    crystal: 0,
                        thunder: 0,
                            metal: 0,
                                feather: 0,
                                    venom: 0,
                                        ice: 0
                                        };
}

const defaultSkills = {
    a1: false,
    a2: false,
    b1: false,
    b2: false,
    c1: false,
    c2: false,
    d1: false,
    d2: false,
    e1: false,
    e2: false
}

const PlayerSchema = new mongoose.Schema({
    user: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true},

    campaign: {type: mongoose.Schema.Types.ObjectId, ref: "Campaign", required: true},

    name: { type: String, required: true },

    class: { type: String, required: true, enum: ["Dareon", "Thoreg", "Mirah", "Karah", "Heleren", "Ljonar"]},

    materials: {
        type: Map,
        of: Number,
        default: defaultMaterials
    },

    plants: {
        type: Map,
        of: Number,
        default: defaultPlants
    },

    elements: {
        type: Map,
        of: Number,
        default: defaultElements
    },

    skills: {
        type: Map,
        of: Boolean,
        default: defaultSkills
    }
})

const Player = mongoose.model('Player', PlayerSchema);

export default Player;

File: MODELS/Entities/Player.cs
using Microsoft.EntityFrameworkCore;

namespace MODELS.Entities;

public class Player
{
    public Guid Id { get; set; }
    public Guid CampaignId { get; set; }
    public Campaign Campaign { get; set; } = null!;

    public string Name { get; set; } = null!;
    public PlayerClass Class { get; set; }

    public List<PlayerSkill> Skills { get; set; } = [];
    public PlayerResources Resources { get; set; } = new();
    public string Notes { get; set; } = string.Empty;
}

public enum PlayerClass
{
    Hunter,
    Guardian,
    Tactician
}

[Owned]
public class PlayerSkill
{
    public SkillCode Skill { get; set; }
    public int Level { get; set; } // 0-2
}

public enum SkillCode
{
    A, B, C, D, E
}

[Owned]
public class PlayerResources
{
    // Plants
    public int Nillea { get; set; }
    public int Tarmaret { get; set; }
    public int Albalacea { get; set; }
    public int Mellis { get; set; }
    public int Anthemon { get; set; }
    public int Saelicornia { get; set; }

    // Materials
    public int Scales { get; set; }
    public int Bones { get; set; }
    public int Blood { get; set; }
    public int Zima { get; set; }
    public int Iride { get; set; }
    public int Kobaureo { get; set; }

    // Elements
    public int Fire { get; set; }
    public int Horn { get; set; }
    public int Coral { get; set; }
    public int Crystal { get; set; }
    public int Thunder { get; set; }
    public int Metal { get; set; }
    public int Feather { get; set; }
    public int Venom { get; set; }
    public int Ice { get; set; }
}